### PR TITLE
refactor(ui): refactor social login hooks

### PR DESCRIPTION
**Pull request by simeng-li**
## Summary

Break social logic into smaller pr. This PR includes the basic social hook design. Will integrate it with ios and android and refine it later.

1. Add a new path /sign-in/callback/:connector to handle the native after the social sign-in callback. Which will render the default sign-in page component and call `signInWithSocial` right after the page is rendered. 
2. useSocial hook
  - load logtoNativeSdk data
  - generate and maintain state includes a uuid, platform data, appid, and postMessage method pass from native.
  - provide the invokeSocialSignIn method to social web ui
  - provide the web and native social SignIn redirect handler for the signing page
  - provide the callback state validation and redirect handler for the social callback page after social signin
  - provide the signInWithSocial handler to sign-in to the logto using returned social status. 

## Linear Issue Reference

log-2166

## Testing

UT case added
will add detailed local test on the next PR

@logto-io/eng

### CODE REVIEW

File: packages/ui/src/utils/index.ts [hunk 1]
```diff
@@ -1,4 +1,14 @@
...
+
+export const parseQueryParameters = (parameters: string | URLSearchParams) => {
+  if (parameters instanceof URLSearchParams) {
+    return Object.fromEntries(parameters.entries());
+  }
+
+  const searchParameters = new URLSearchParams(parameters);
+
+  return Object.fromEntries(searchParameters.entries());
+};
```
Comment (IceHe) [inline packages/ui/src/utils/index.ts:14]: simplier?
```ts
export const parseQueryParameters = (parameters: string | URLSearchParams) => {
  const searchParameters =
    parameters instanceof URLSearchParams ? parameters : new URLSearchParams(parameters);

  return Object.fromEntries(searchParameters.entries());
};
```

File: packages/ui/src/hooks/use-social.ts [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+  const state: State = { uunid, platform, callbackUriScheme };
+
+  return Buffer.from(JSON.stringify(state)).toString('base64');
+};
+
+export const decodeState = (state: string) => {
+  try {
+    const value = JSON.parse(Buffer.from(state, 'base64').toString()) as State;
+
+    return value;
```
Comment (IceHe) [inline packages/ui/src/hooks/use-social.ts:40]: nit: just inline `value`?

File: packages/ui/src/hooks/use-social.ts [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+        return;
+      }
+
+      if (!callbackUriScheme) {
+        // TODO: native callbackUriScheme not found error message
+        return;
+      }
+
+      window.location.assign(`${callbackUriScheme}${connectorId}${window.location.search}`);
```
Comment (xiaoyijun) [inline packages/ui/src/hooks/use-social.ts:127]: per offline discussion, maybe hard code it `logto-callback://${appId}/web?${search}`.
https://silverhand-io.slack.com/archives/C02A8G4HVAM/p1650343181926779

File: packages/ui/src/components/Toast/index.module.scss [hunk 4]
```diff
@@ -11,16 +11,17 @@
...
     font: var(--font-body-medium);
     color: var(--color-font-toast-text);
     border-radius: _.unit(2);
-    max-width: none;
     background: var(--color-dark-background);
     min-width: _.unit(25);
     text-align: center;
     opacity: 0%;
     transition: opacity 0.3s ease-in-out;
+    word-break: break-all;
```
Comment (demonzoo) [inline packages/ui/src/components/Toast/index.module.scss:24]: Would `break-word` be better?

File: packages/ui/src/hooks/use-social.ts [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+import { invokeSocialSignIn, signInWithSocial } from '@/apis/social';
+import { generateRandomString, parseQueryParameters } from '@/utils';
+
+import PageContext from './page-context';
+import useApi from './use-api';
+
+/**
+ * Social Connector State Utility Methods
+ * @param state
+ * @param state.uunid - unique id
```
Comment (demonzoo) [inline packages/ui/src/hooks/use-social.ts:13]: uuid?

File: packages/ui/src/containers/SocialSignIn/SecondarySocialSignIn.test.tsx [hunk 6]
```diff
@@ -1,12 +1,40 @@
...
+  const signInWithSocialSpy = jest
+    .spyOn(socialSignInApi, 'signInWithSocial')
+    .mockResolvedValue({ redirectTo: `${mockOrigin}/callback` });
+
+  beforeEach(() => {
+    /* eslint-disable @silverhand/fp/no-mutation */
+    // @ts-expect-error mock global object
+    globalThis.logtoNativeSdk = {
+      platform: 'web',
+      getPostMessage: jest.fn(() => jest.fn()),
```
Comment (wangsijie) [inline packages/ui/src/containers/SocialSignIn/SecondarySocialSignIn.test.tsx:28]: is this on purpose?

File: packages/ui/src/hooks/use-social.ts [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+const storageKeyPrefix = 'social_auth_state';
+
+export const generateState = () => {
+  const uunid = generateRandomString();
+  const platform = logtoNativeSdk?.platform ?? 'web';
+  const callbackUriScheme = logtoNativeSdk?.callbackUriScheme;
+
+  const state: State = { uunid, platform, callbackUriScheme };
+
+  return Buffer.from(JSON.stringify(state)).toString('base64');
```
Comment (demonzoo) [inline packages/ui/src/hooks/use-social.ts:33]: Better wrap this in a try-catch?

File: packages/ui/src/hooks/use-social.ts [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+type State = {
+  uunid: string;
+  platform: 'web' | 'ios' | 'android';
+  callbackUriScheme?: string;
+};
+
+const storageKeyPrefix = 'social_auth_state';
+
+export const generateState = () => {
+  const uunid = generateRandomString();
```
Comment (wangsijie) [inline packages/ui/src/hooks/use-social.ts:27]: uuid?

File: packages/ui/src/hooks/use-social.ts (hunk 7 above)
Comment (wangsijie) [inline packages/ui/src/hooks/use-social.ts:33]: can we use native `btoa` function?

File: packages/ui/src/hooks/use-social.ts [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+    const value = JSON.parse(Buffer.from(state, 'base64').toString()) as State;
+
+    return value;
+  } catch {}
+};
+
+export const stateValidation = (state: string, connectorId: string) => {
+  const stateStorage = sessionStorage.getItem(`${storageKeyPrefix}:${connectorId}`);
+
+  return Boolean(stateStorage) && stateStorage === state;
```
Comment (demonzoo) [inline packages/ui/src/hooks/use-social.ts:47]: `Boolean(stateStorage)` can be omitted, as the `state` is string type, which has no overlap with a possible `null` value

File: packages/ui/src/hooks/use-social.ts [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+    }
+  }, [parameters.connector, socialCallbackHandler]);
+
+  // Monitor Native Error Message
+  useEffect(() => {
+    window.addEventListener('message', function (event) {
+      if (event.origin === window.location.origin) {
+        setToast(JSON.stringify(event.data));
+      }
+    });
```
Comment (demonzoo) [inline packages/ui/src/hooks/use-social.ts:175]: nit: use arrow function

File: packages/ui/src/include.d/global.d.ts [hunk 11]
```diff
@@ -0,0 +1,10 @@
+// Logto Native SDK
+
+type LogtoNativeSdkInfo = {
+  platform: 'ios' | 'android';
+  callbackUriScheme: string;
+  getPostMessage: () => (data: { callbackUri?: string; redirectTo?: string }) => void;
```
Comment (demonzoo) [inline packages/ui/src/include.d/global.d.ts:6]: Does this funtion return a function?

File: packages/ui/src/containers/SocialSignIn/SecondarySocialSignIn.test.tsx (hunk 6 above)
Comment (simeng-li) [reply packages/ui/src/containers/SocialSignIn/SecondarySocialSignIn.test.tsx:28]: `jest.fn(() => jest.fn())?`  Yes, its function returns a function

File: packages/ui/src/hooks/use-social.ts (hunk 7 above)
Comment (simeng-li) [reply packages/ui/src/hooks/use-social.ts:33]: image

technically browser is fine? I'll update then

File: packages/ui/src/include.d/global.d.ts (hunk 11 above)
Comment (simeng-li) [reply packages/ui/src/include.d/global.d.ts:6]: yes indeed.

File: packages/ui/src/components/Toast/index.module.scss (hunk 4 above)
Comment (simeng-li) [reply packages/ui/src/components/Toast/index.module.scss]: updated

File: packages/ui/src/hooks/use-social.ts (hunk 5 above)
Comment (simeng-li) [reply packages/ui/src/hooks/use-social.ts]: updated

File: packages/ui/src/hooks/use-social.ts (hunk 3 above)
Comment (simeng-li) [reply packages/ui/src/hooks/use-social.ts]: updated

File: packages/ui/src/hooks/use-social.ts (hunk 2 above)
Comment (simeng-li) [reply packages/ui/src/hooks/use-social.ts]: updated

File: packages/ui/src/hooks/use-social.ts (hunk 8 above)
Comment (simeng-li) [reply packages/ui/src/hooks/use-social.ts]: updated

File: packages/ui/src/hooks/use-social.ts (hunk 7 above)
Comment (simeng-li) [reply packages/ui/src/hooks/use-social.ts]: updated

File: packages/ui/src/hooks/use-social.ts (hunk 9 above)
Comment (simeng-li) [reply packages/ui/src/hooks/use-social.ts]: updated

File: packages/ui/src/hooks/use-social.ts (hunk 10 above)
Comment (simeng-li) [reply packages/ui/src/hooks/use-social.ts]: updated also add the return removeEvent

**Review by demonzoo — APPROVED** [review]

**Pull request merged**